### PR TITLE
Perform part of the pipeline using Flow

### PR DESCRIPTION
**Pull request by msramos**
Leverage the most expensive part of the processing pipeline to Flow

## Benchmark

I ran both benchmarks with a small Phoenix project. The results are:

*Without Flow*
```
Name                 ips        average  deviation         median         99th %
find_typos        0.0116       1.43 min     ±0.00%       1.43 min       1.43 min

Memory usage statistics:

Name          Memory usage
find_typos         1.28 GB
```

*With Flow*
```
Name                 ips        average  deviation         median         99th %
find_typos        0.0120       1.38 min     ±0.00%       1.38 min       1.38 min

Memory usage statistics:

Name          Memory usage
find_typos        46.34 MB
```

### Summary
```
Time performance: -4.9%
Memory usage:    -96.3%
```

### CODE REVIEW

File: lib/typo_killer/words_parser.ex [hunk 1]
```diff
@@ -17,7 +17,12 @@ defmodule TypoKiller.WordsParser do
     |> MapSet.new()
   end
 
-  @spec find_words(file :: String.t()) :: list()
```
Comment (samuelpordeus) [inline lib/typo_killer/words_parser.ex:20]: I think you accidentally deleted this spec here

Comment (msramos) [reply lib/typo_killer/words_parser.ex:20]: Fixed

File: lib/typo_killer/finder.ex [hunk 2]
```diff
@@ -12,8 +12,10 @@ defmodule TypoKiller.Finder do
 
   def find_typos({words, dictionary}) do
     words
-    |> Enum.map(&Task.async(fn -> calculate_distance(&1, dictionary) end))
-    |> Enum.flat_map(&Task.await(&1, :infinity))
+    |> Flow.from_enumerable()
+    |> Flow.map(fn word -> calculate_distance(word, dictionary) end)
+    |> Flow.reduce(fn -> MapSet.new() end, fn a, b -> MapSet.union(MapSet.new(a), b) end)
+    |> Enum.to_list()
```
Comment (samuelpordeus) [inline lib/typo_killer/finder.ex]: why is this line necessary?

File: lib/typo_killer/finder.ex [hunk 3]
```diff
@@ -12,8 +12,10 @@ defmodule TypoKiller.Finder do
 
   def find_typos({words, dictionary}) do
     words
-    |> Enum.map(&Task.async(fn -> calculate_distance(&1, dictionary) end))
-    |> Enum.flat_map(&Task.await(&1, :infinity))
+    |> Flow.from_enumerable()
+    |> Flow.map(fn word -> calculate_distance(word, dictionary) end)
+    |> Flow.reduce(fn -> MapSet.new() end, fn a, b -> MapSet.union(MapSet.new(a), b) end)
```
Comment (samuelpordeus) [inline lib/typo_killer/finder.ex]:
```suggestion
    |> Flow.reduce(fn -> MapSet.new() end, fn word, words_set -> MapSet.union(MapSet.new(word), words_set) end)
```

I think something more meaningful for `a` and `b` here would be nice :thinking:

Comment (samuelpordeus) [reply lib/typo_killer/finder.ex]: (not that x is better, but I lack creativity right now :laughing:)

Comment (samuelpordeus) [reply lib/typo_killer/finder.ex]: maybe `word` and `words_set`?

**Comment by mracos** [comment]
holy cow what a beautiful PR, I'm your number one:
giphub

**Comment by hugobarauna** [comment]
22 lines added to get a 96% improvement in memory consumption, HOLY MOLY!  👏 🙀 👏 🙀 👏 🙀

File: lib/typo_killer/finder.ex [hunk 4]
```diff
@@ -11,12 +11,18 @@ defmodule TypoKiller.Finder do
 
   def find_typos({words, dictionary}) do
     words
-    |> Enum.map(&Task.async(fn -> calculate_distance(&1, dictionary) end))
-    |> Enum.flat_map(&Task.await(&1, :infinity))
+    |> Flow.from_enumerable()
```
Comment (hugobarauna) [inline lib/typo_killer/finder.ex]: I love how simple it is was to move from serialized processing to concurrent processing. 😍

File: lib/typo_killer/finder.ex [hunk 5]
```diff
@@ -11,12 +11,18 @@ defmodule TypoKiller.Finder do
...
+    |> Flow.map(fn word -> calculate_distance(word, dictionary) end)
+    |> Flow.reduce(fn -> MapSet.new() end, &merge_partial_result/2)
```
Comment (hugobarauna) [inline lib/typo_killer/finder.ex:16]: I love how simple it is was to move from serialized processing to concurrent processing. 😍
